include/machine/fpu.h

```diff
@@ -16,43 +16,26 @@
 /* Perform any actions required for the deletion of the given thread. */
 void fpuThreadDelete(tcb_t *thread);
 
-/* Handle an FPU exception. */
-exception_t handleFPUFault(void);
-
-void switchLocalFpuOwner(user_fpu_state_t *new_owner);
+void switchLocalFpuOwner(tcb_t *new_owner);
 
 /* Switch the current owner of the FPU state on the core specified by 'cpu'. */
-void switchFpuOwner(user_fpu_state_t *new_owner, word_t cpu);
+void switchFpuOwner(tcb_t *new_owner, word_t cpu);
 
 /* Returns whether or not the passed thread is using the current active fpu state */
 static inline bool_t nativeThreadUsingFPU(tcb_t *thread)
 {
-    return &thread->tcbArch.tcbContext.fpuState ==
-           NODE_STATE_ON_CORE(ksActiveFPUState, thread->tcbAffinity);
+    return thread == NODE_STATE_ON_CORE(ksActiveFPUThread, thread->tcbAffinity);
 }
 
+/* Called without global lock held! */
 static inline void FORCE_INLINE lazyFPURestore(tcb_t *thread)
 {
```

Comment:
Unfortunately, now that we're actually modelling lazy FPU switching, this being called from `restore_user_context` and `fastpath_restore` is going to cause major difficulties. The way those functions are modelled in the specifications are very unique, and won't be able to handle `lazyFPURestore` with its conditionals and reading and writing to kernel state.

Would it be possible to call this in the scheduler instead, as part of `switchToThread`, in roughly the same way that VCPU state being switched is handled? That would be much better from the verification perspective.

Reply:
To also add to this: restoring thread user context used to be unconditional and did not modify kernel state, which is why in verification we could sweep it under the rug as part of the kernel enter-exit procedure. As Corey says, `lazyFPURestore` doesn't meet the criteria for safe wrap-and-forget, so `switchToThread` looks like a good place (directly before or after `Arch_switchToThread`).

Reply:
I suspected this would be a problem. I'll have a look at moving it.